pkg/agent/nodeportlocal/k8s/annotations.go

```diff
@@ -38,9 +38,10 @@ const (
 
 // NPLAnnotation is the structure used for setting NodePortLocal annotation on the Pods.
 type NPLAnnotation struct {
-	PodPort  int    `json:"podPort"`
-	NodeIP   string `json:"nodeIP"`
-	NodePort int    `json:"nodePort"`
+	PodPort   int      `json:"podPort"`
+	NodeIP    string   `json:"nodeIP"`
+	NodePort  int      `json:"nodePort"`
+	Protocols []string `json:"protocols"`
```

Comment:
I am confused:
- on the one hand, you are using a slice of protocols here which would imply that the same NPL annotation can be used for multiple protocols
- on the other hand, it seems that you only ever include one protocol in this slice:

```
			nplAnnotationsRequiredMap[nodePort] = NPLAnnotation{
				PodPort:   port,
				NodeIP:    pod.Status.HostIP,
				NodePort:  nodePort,
				Protocols: []string{protocol},
			}
```

Maybe this shouldn't be a slice?

Or maybe we should consider using the same NodePort when the target port / PodPort is used for both "UDP" and "TCP", even though I don't think this is done for K8s NodePort Services. Have you given any thought to this?

Reply:
The single protocol in the slice seems to be a misunderstanding. That part of the code only initially fills in the slice with a single protocol, later the same NodePort is used for another protocol, in which case I do append the `Protocols` slice. Then the conditional right before that would be used.
```
                if val, ok := nplAnnotationsRequiredMap[nodePort]; ok {
 			val.Protocols = append(val.Protocols, protocol)
 			nplAnnotationsRequiredMap[nodePort] = val
 		}
```

The annotation represents the Protocol as a slice per NodePort to maintain NodePort uniqueness in the annotation. 
The PortTable struct used to populate iptable rules is kept as is, to keep changes minimal there.

It is intended as part of this PR that we use the same NodePort when the pod port is used for both TCP and UDP. The e2e test has a sample usecase and intended behaviour included.

Reply:
Thanks for clarifying. Could you add unit tests to pkg/agent/nodeportlocal/npl_agent_test.go as well for this case? Just looking at the code, it's not obvious to me that the following case is supported:

* Pod PA is created and serves on port 80 for both UDP and TCP
* Pod PB is created and serves on port 80 for UDP only
* NPL Service SA is created and targets Pod PA (TCP only)
  - at this stage Pod PA should be exposed on NodePort 61000 (TCP only)
* NPL Service SB is created and targets Pod PB (UDP only)
  - at this stage Pod PB should be exposed on NodePort 61001 (UDP only)
* NPL Service SB is modified to target both Pod PB and Pods PA (UDP only)
  - at this stage PodPA should be exposed on NodePort 61000 (TCP & UDP)

Thanks!

Reply:
@antoninbas My latest commit makes sure that a given NodePort is unique to a PodIP/Pod. Earlier, it was possible that a given NodePort could be shared among Pods because of different Protocols, and such a situation could arise
```
PA: [{61000, [tcp]}, {61001, [udp]}]
PB: [{61000, [udp]}, {61001, [tcp]}]
```
This has been avoided. I think the proper way would be to maintain a unique endpoint `(NodeIP:NodePort)` per Pod.
I have also added a unit test for the same, in the latest commit.

Reply:
Resolving this since you addressed the issue and added a test